src/algebra/quadratic_discriminant.lean

```diff
@@ -39,46 +41,51 @@ variables {R : Type*}
 /-- Discriminant of a quadratic -/
 def discrim [ring R] (a b c : R) : R := b^2 - 4 * a * c
 
-variables [comm_ring R] [is_domain R] {a b c : R}
+@[simp] lemma discrim_neg [ring R] (a b c : R) : discrim (-a) (-b) (-c) = discrim a b c :=
+by simp [discrim]
+
+variables [comm_ring R] {a b c : R}
+
+lemma discrim_eq_sq_of_quadratic_eq_zero {x : R} (h : a * x * x + b * x + c = 0) :
+  discrim a b c = (2 * a * x + b) ^ 2 :=
+begin
+  rw [discrim],
+  linear_combination -4 * a * h
+end
 
 /--
 A quadratic has roots if and only if its discriminant equals some square.
 -/
-lemma quadratic_eq_zero_iff_discrim_eq_sq (h2 : (2 : R) ≠ 0) (ha : a ≠ 0) (x : R) :
+lemma quadratic_eq_zero_iff_discrim_eq_sq [ne_zero (2 : R)] [no_zero_divisors R]
+  (ha : a ≠ 0) {x : R} :
   a * x * x + b * x + c = 0 ↔ discrim a b c = (2 * a * x + b) ^ 2 :=
 begin
-  dsimp [discrim] at *,
-  split,
-  { assume h,
-    linear_combination -4 * a * h },
-  { assume h,
-    have ha : 2 * 2 * a ≠ 0 := mul_ne_zero (mul_ne_zero h2 h2) ha,
-    apply mul_left_cancel₀ ha,
-    linear_combination -h }
+  refine ⟨discrim_eq_sq_of_quadratic_eq_zero, λ h, _⟩,
+  rw [discrim] at h,
+  have ha : 2 * 2 * a ≠ 0 := mul_ne_zero (mul_ne_zero (ne_zero.ne _) (ne_zero.ne _)) ha,
+  apply mul_left_cancel₀ ha,
+  linear_combination -h
 end
 
 /-- A quadratic has no root if its discriminant has no square root. -/
-lemma quadratic_ne_zero_of_discrim_ne_sq (h2 : (2 : R) ≠ 0) (ha : a ≠ 0)
-  (h : ∀ s : R, discrim a b c ≠ s * s) (x : R) :
+lemma quadratic_ne_zero_of_discrim_ne_sq (h : ∀ s : R, discrim a b c ≠ s * s) (x : R) :
```

Comment:
Given this uses `sq` in the name, should we just change the statement to use `s ^ 2`? That would also save you a rewrite.